### ISSUE TITLE
IndexDB still broken on Firefox 71

### ISSUE THREAD

**Issue by dali99** [NixOS/nixpkgs#76899]
**Describe the bug**
in https://github.com/NixOS/nixpkgs/commit/6c793254dacfe35f730abb8b9be1e35bf1844238 indexDB was supposedly fixed. But in riot.im E2E encryption is broken with:
```
IndexedDBStore failure, degrading to MemoryStore [object Object]: ""
```

**To Reproduce**
1. open a riot-im client (riot.im/app)
2. start an E2E encrypted chat with someone (for example me! [@dandellion:dodsorf.as](https://matrix.to/#/@dandellion:dodsorf.as))
3. Fail at reading the messages :S

**Additional context**
https://github.com/vector-im/riot-web/issues/11711 has a few nixos users confirming this
https://github.com/vector-im/riot-web/issues/11606 more information

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.91, NixOS, 19.09.1690.0d9055a2ac2 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(dan): `"home-manager-19.09, nixos-19.09.1693.eab4ee0c27c, nixos-unstable-20.03pre206632.b0bbacb5213"`
 - channels(root): `"nixos-19.09.1693.eab4ee0c27c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: firefox
```

@maralorn 

@andir

**Comment by andir** [comment]
Related bugzilla entry: https://bugzilla.mozilla.org/show_bug.cgi?id=1601707
Fedora bugfix: https://src.fedoraproject.org/rpms/firefox/c/0f0618f13f764427bd55921e8a3cb7469984391b?branch=master
Debian bugfix: https://salsa.debian.org/mozilla-team/firefox/commit/e888ce1578d68116949782035f8149cdbc5e832c

Upstream fix: https://phabricator.services.mozilla.com/D56873


Looking at the comments in the linked reports Fedora and Debian fixed it. The changes they applied are somewhat similar to what @fpletz came up with. The upstream patch looks much better.

@dali99 @maralorn: are either of you able to test the above linked upstream patch on 19.09? I do not have any use-case that requries that triggers said bugs :/


We should probably come up with a test case for this and add it to our test suite.

FFX72 was just branched off so that is what I'll work on during the weekend. Supposedly it already has the fix linked above.

**Comment by thoferon** [comment]
For people in need of a workaround in the meantime, `nixpkgs.firefox-beta-bin` (72.0b4) fixed the issue for me.

**Comment by dali99** [comment]
> @dali99 @maralorn: are either of you able to test the above linked upstream patch on 19.09? I do not have any use-case that requries that triggers said bugs :/

I cloned nixpkgs and applied this diff to the nixos 19.09 branch before running `nix-build $NIXPKGS -A firefox`

```
diff --git a/pkgs/applications/networking/browsers/firefox/common.nix b/pkgs/applications/networking/browsers/firefox/common.nix
index a6d57ae8edd..2609f154b27 100644
--- a/pkgs/applications/networking/browsers/firefox/common.nix
+++ b/pkgs/applications/networking/browsers/firefox/common.nix
@@ -116,7 +116,12 @@ let
       url = "https://raw.githubusercontent.com/archlinuxarm/PKGBUILDs/09c7fa0dc1d87922e3b464c0fa084df1227fca79/extra/firefox/build-arm-libopus.patch";
       sha256 = "1zg56v3lc346fkzcjjx21vjip2s9hb2xw4pvza1dsfdnhsnzppfp";
     })
-  ] ++ lib.optional (lib.versionAtLeast ffversion "71") ./fix-ff71-lto.patch
+  ]
+  ++ lib.optional (lib.versionAtLeast ffversion "71")
+    (fetchpatch {
+      url = "https://phabricator.services.mozilla.com/D56873?download=true";
+      sha256 = "02dwd8mqx9rbm2d36ac2w1chmycygmcd5g1plbgbj577ghkv1fj6";
+    })
   ++ patches;
 
   nss_pkg = if lib.versionAtLeast ffversion "71" then nss_3_47_1 else nss;
```

But running the resulting binary still makes encryption not work. Though I have a sneaking suspicion this is for an unrelated issue in how I tested this. (I think the profile directory is wrong (unwritable)) so a pointer to how I'm supposed to test firefox builds would be much appreciated!

> For people in need of a workaround in the meantime, nixpkgs.firefox-beta-bin (72.0b4) fixed the issue for me.

Just to confirm this is compatible with your profile right?

**Comment by fpletz** [comment]
This might be related to the recent GCC9 bump since Fedora also had issues we didn't encounter with GCC8.

**Comment by thoferon** [comment]
> Just to confirm this is compatible with your profile right?

Yes, my profile was still there after the upgrade.

**Comment by dali99** [comment]
I ran `home-manager -I nixpkgs=$NIXPKGS switch` with the new patch and it still failed.

So the upstream patch is not enough to resolve this.

**Comment by maralorn** [comment]
@andir I am not sure what to test here. The patch @dali99 applied to nixpkgs is the one you had send me before, which didn‘t help.

If this is fixed in firefox 72 I am okay with waiting. Of course the workaround @thoferon could just work because the bin release does not have the same gcc problems.

**Comment by andir** [comment]
There seems to be a solution for the matrix issue that is described here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1601707#c40

The correct patch is supposedly this one: https://phabricator.services.mozilla.com/D56876
The other patches (for newer versions of Firefox) look very similar but were missing one case and thus didn't fix the issue but probably the crash(?).

I am waiting for feedback from @maralorn if that patch works as expected.

**Comment by fpletz** [comment]
Fixed by #77331 & #77292.
